### ISSUE TITLE
Random BertForClassification inference scores for a single output feature

### ISSUE THREAD

**Issue by AMR-KELEG** [huggingface/hub-docs#738]
**Is your feature request related to a problem? Please describe.**
I have a fine-tuned BERT-based model with a Linear classification Head on top of the sentence encoding. The head has a single output cell (`(classifier): Linear(in_features=768, out_features=1, bias=True)`). This setup is basically equivalent to having a regression head on top of BERT. The model works as expected locally.  However, when I try to test the Inference API on the website, I get the following message `⚠️ This model could not be loaded by the inference API. ⚠️`. When I input a sentence, the site still generates an output, but the output seems to be random.

**Describe the solution you'd like**
Allow the site to generate row scores if the classification head has a single output neuron.

**Describe alternatives you've considered**
NA

**Additional context**

<img width="532" alt="Screenshot 2023-04-12 at 01 22 57" src="https://user-images.githubusercontent.com/8365743/231309676-fbdeb9fe-e37a-412e-ad6e-269dcb49475c.png">
<img width="532" alt="Screenshot 2023-04-12 at 01 23 35" src="https://user-images.githubusercontent.com/8365743/231309688-fa1d0279-b940-42f3-b676-9d75128db51b.png">

Output from the locally-loaded model: `SequenceClassifierOutput(loss=None, logits=tensor([[0.0263]], grad_fn=<AddmmBackward0>), hidden_states=None, attentions=None)
`

**Comment by osanseviero** [comment]
Hi there! What's the repository id for this project? Would love to take a look to understand the issue

**Comment by AMR-KELEG** [comment]
Hi @osanseviero 
Apologies for my late reply.
I uploaded a new toy model to demonstrate the issue.

Model: https://huggingface.co/AMR-KELEG/toy_regression_model
Pretraining Code:
```
import torch
from transformers import BertForSequenceClassification, AutoTokenizer

# Model pretraining
model_name = "bert-base-uncased"

tokenizer = AutoTokenizer.from_pretrained(model_name)
model = BertForSequenceClassification.from_pretrained(
    model_name, num_labels=1
)

model.save_pretrained("toy_regression_model")
tokenizer.save_pretrained("toy_regression_model")
```

Inference Code:
```
import torch
from transformers import BertForSequenceClassification, AutoTokenizer

# Model Loading
model_name = "AMR-KELEG/toy_regression_model"

loaded_model = BertForSequenceClassification.from_pretrained(model_name)
load_tokenizer = AutoTokenizer.from_pretrained(model_name)

loaded_model(**load_tokenizer(["I like you. I love you"], return_tensors='pt'))
```

Local output:
```
SequenceClassifierOutput(loss=None, logits=tensor([[0.3264]], grad_fn=<AddmmBackward0>), hidden_states=None, attentions=None)
```

Hosted inference API's output:
```
[
  [
    {
      "label": "LABEL_0",
      "score": 0.5808932185173035
    }
  ]
]
```

**Comment by osanseviero** [comment]
I just tried with `transformers` pipeline and it seems to match the widget (0.58). 

```python
from transformers import pipeline

pipe = pipeline("text-classification", model="AMR-KELEG/toy_regression_model")
pipe("I like you. I love you")
# [{'label': 'LABEL_0', 'score': 0.5808932185173035}]
```

It's true though that

```
import torch
from transformers import BertForSequenceClassification, AutoTokenizer

# Model Loading
model_name = "AMR-KELEG/toy_regression_model"

loaded_model = BertForSequenceClassification.from_pretrained(model_name)
load_tokenizer = AutoTokenizer.from_pretrained(model_name)

loaded_model(**load_tokenizer("I like you. I love you", return_tensors='pt'))
```

gives an inconsistent output, maybe @Narsil has an idea here

**Comment by LysandreJik** [comment]
Hey! The difference between is the two is that by using the model/tokenizer like this, you're retrieving the raw logits. 

The pipeline post-processes the output so that it is interpretable. According to the classification type: it can pass the logit through a sigmoid (in case of a multi-label classification), a softmax (in case of a single-label classification), or none at all if specified as such.

Your model has `loaded_model.config.num_labels == 1`, therefore using the sigmoid method to post-process the model's output.

Here's how to obtain the same result:

```py
In [1]: import torch
    ...: from transformers import BertForSequenceClassification, AutoTokenizer
    ...: 
    ...: # Model Loading
    ...: model_name = "AMR-KELEG/toy_regression_model"
    ...: 
    ...: loaded_model = BertForSequenceClassification.from_pretrained(model_name)
    ...: load_tokenizer = AutoTokenizer.from_pretrained(model_name)
    ...: 
    ...: outputs = loaded_model(**load_tokenizer("I like you. I love you", return_tensors='pt'))
    ...: 
    ...: def sigmoid(_outputs):
    ...:     return 1.0 / (1.0 + np.exp(-_outputs))
    ...: 
    ...: sigmoid(outputs.logits.detach().numpy())


Out[1]: array([[0.5808932]], dtype=float32)
```

**Comment by AMR-KELEG** [comment]
Thanks for investigating.
Since the model is trained without a sigmoid function, is there a way to disable it on the inference API as well? (i.e.: output the raw logits without applying the sigmoid function)

**Comment by osanseviero** [comment]
No, in the context of classification, it makes sense for the widget to output the post-sigmoid result to be interpreted.

**Comment by osanseviero** [comment]
No, in the context of classification, it makes sense for the widget to output the post-sigmoid result to be interpreted.

**Comment by AMR-KELEG** [comment]
Hi @osanseviero 
I thought of multiple hacks, but failed to find any solution for having my model produce the correct scores through the HF Inference API. My model is basically generating a single score in the range [0, 1]. It's basically a regression model, so I didn't want to have a softmax function on top of it.

The work-arounds I can now think of is:
1) Finding a way to pass the `function_to_apply="none"` to the constructor of the pipeline on HF's Inference API through the config.json file. I am expecting that this solution is infeasible, though.
2) Adding an extra dummy label on top of the model, and apply an inverse softmax function so that the original score for the first label is the score I am interested in, while the other score is always set to a small value.

Could you please advise if there is a clean way for having the continuous score in range [0,1] being shown through the HF inference API without using lots (hopefully none) of these hacks?

Thanks!

**Comment by osanseviero** [comment]
Hi there! You can specify extra parameters as this: 

```python
import requests
API_URL = "https://api-inference.huggingface.co/models/vectara/hallucination_evaluation_model"
def query(payload):
    response = requests.post(API_URL, json=payload)
    print(payload)
    return response.json()
data = query({"inputs": "I like you. I love you", "parameters": {"function_to_apply": "none"}})
```
